scipy/stats/stats.py

```diff
@@ -4275,6 +4276,164 @@ def mannwhitneyu(x, y, use_continuity=True):
+    possible combinations is <100000. The default behavior may be overridden
+    with the use_exact flag.
+
+    If an exact test is not performed, the U-distribution is approximated as a
+    normal distribution.
+
+    This test corrects for ties and by default uses a continuity correction
+    when approximating the U statistic distribution.
+
+    The reported p-value is for a two-sided hypothesis. To get the one-sided
+    p-value set alternative to 'greater' or 'less' (default is 'two-sided').
+
+    For Mann-Whitney-Wilcoxon tests, the reported U statistic is the U used to
+    test the hypothesis. The u1 and u2 statistics are returned as well,
+    corresponding to x and y, respectively.
+
+
+    .. versionadded:: 0.17.0
+
+    References
+    ----------
+    .. [1] H.B. Mann and D.R. Whitney, "On a test of whether one of two random
+           variables is stochastically larger than the other", The Annals of
+           Mathematical Statistics, Vol. 18, pp. 50-60, 1947.
+    .. [2] http://en.wikipedia.org/wiki/Mann-Whitney_U_test
+    """
+    x = asarray(x)
+    y = asarray(y)
+    if len(np.shape(x)) > 1 or len(np.shape(y)) > 1:
+        raise ValueError('Expected 1-d arrays.')
+    n1 = len(x)
+    n2 = len(y)
+    if exact == 'auto':
+        exact = ((n1 < 10 or n2 < 10) and n1 + n2 < 100000
+                 and -np.log(n1 + n2 + 1) - special.betaln(n1 + 1, n2 + 1) < np.log(100000))
+    ranked = rankdata(np.concatenate((x,y)))
+    rankx = ranked[0:n1]       # get the x-ranks
+    T = tiecorrect(ranked)
+    if T == 0:
+        raise ValueError('All numbers are identical')
+    if alternative not in ('two-sided', 'less', 'greater'):
+        raise AttributeError("Alternative should be one of: "
+                             "'two-sided', 'less', or 'greater'")
+    if exact:
+        a = list(range(n1, n1+n2))
+        u = [0]
+        while sum(a) != sum(range(n2)):   # When in leftmost position, a == list(range(n2))
+            # Do the shift operation
+            i = 0
+            while a[i] == i:
+                i += 1
+            a[i] -= 1
+            j = i - 1
+            while j >= 0:
+                a[j] = a[i] - (i - j)
+                j -= 1
+            # count(a < a1) = U2
+            u1 = 0
+            for i, x in enumerate(a):
+                u1 += n1 - x + i
+            u2 = n1*n2 - u1
+            # store min U value to array
+            if alternative == 'two-sided':
+                u.append(min(u1, u2))
+            else:
+                u.append(u1)
+        u1 = 0
+        u = np.array(u)
+        # for i, x in enumerate(sorted(rankx)):
+        #     u1 += x - 1 - i
+        u1 = rankx.sum() - n1*(n1+1)/2
+        u2 = n1 * n2 - u1
+        if alternative == 'two-sided':
+            smallu = min(u1, u2)
+        elif alternative == 'greater':
+            smallu = u1
+        else:
+            smallu = u2
+        p = sum(u <= smallu) / len(u)
```

Comment:
Nitpick: `u` is an array, so can do `u[u <= smallu].sum() / u.size`

Reply:
Also, this seems to return the complement of the desired p value for the one-sided alternatives. Should add a test that the exact and asymptotic test agree for large inputs.

```
x = [7, 5, 6, 4, 12]
y = [3, 6, 4, 2, 1]
for alternative in ('less', 'greater', 'two-sided'):
    res0 = mannwhitneyu(x, y, alternative=alternative)
    res1 = mann_whitney_u(x, y, alternative=alternative, exact=False)
    res2 = mann_whitney_u(x, y, alternative=alternative, exact=True)
    print(res0.pvalue, res1.pvalue, res2.pvalue)
```
```
0.982210583380203 0.982210583380203 0.027777777777777776
0.029276315078413294 0.029276315078413294 0.9841269841269841
0.05855263015682659 0.05855263015682659 0.05555555555555555
```